cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,311 @@
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Polls the JobQueue, checks the jobs for constraint satisfaction, and launches them.
+ */
+class ConstraintCheckerService extends AbstractService {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintCheckerService.class);
+
+  private final Transactional transactional;
+  private final DatasetFramework datasetFramework;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+
+  @Inject
+  ConstraintCheckerService(Store store,
+                           ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                           NamespaceQueryAdmin namespaceQueryAdmin,
+                           CConfiguration cConf,
+                           DatasetFramework datasetFramework,
+                           TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+  }
+
+  @Override
+  protected void doStart() {
+    if (!isRunning()) {
+      return;
+    }
+    LOG.info("Starting ConstraintCheckerService.");
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("constraint-checker-task").build()));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+
+    for (int partition = 0; partition < getNumPartitions(); partition++) {
+      taskExecutorService.submit(new ConstraintCheckerThread(partition));
+    }
+    LOG.info("Started ConstraintCheckerService.");
+  }
+
+  @Override
+  protected void doStop() {
+    LOG.info("Stopping ConstraintCheckerService.");
+    try {
+      if (!taskExecutorService.awaitTermination(5, TimeUnit.SECONDS)) {
```

Comment:
This for sure will return `false`. You have to call `shutdown` or `shutdownNow` first before calling `awaitTermination`.